src/core/transport/chttp2_transport.c

```diff
@@ -997,15 +995,29 @@ static void perform_transport_op(grpc_exec_ctx *exec_ctx, grpc_transport *gt,
     close_transport_locked(exec_ctx, t);
   }
 
-  unlock(exec_ctx, t);
-
   if (close_transport) {
-    lock(t);
     close_transport_locked(exec_ctx, t);
-    unlock(exec_ctx, t);
   }
 }
 
+static void perform_transport_op(grpc_exec_ctx *exec_ctx, grpc_transport *gt,
+                                 grpc_transport_op *op) {
+  grpc_chttp2_transport *t = (grpc_chttp2_transport *)gt;
+
+  lock(t);
+
+  /* Let's be overly cautious: don't change any state while we're parsing */
+  if (t->parsing_active) {
+    GPR_ASSERT(t->post_parsing_op == NULL);
+    t->post_parsing_op = gpr_malloc(sizeof(*op));
+    memcpy(t->post_parsing_op, op, sizeof(*op));
+  } else {
```

Comment:
I don't think there's any case where we'd come down this path more than
once during the same parse. If it ever became necessary we could add the
necessary extra machinery to make it work (either by merging ops or
creating a to-do list).

Either seems like adding excess machinery too early however, and so I'm
reluctant to do it until a use case arrives.

On Thu, Mar 3, 2016, 5:16 PM Vijay Pai <notifications@github.com> wrote:

> In src/core/transport/chttp2_transport.c
> <https://github.com/grpc/grpc/pull/5560#discussion_r54977577>:
>
> >    }
> >  }
> >
> > +static void perform_transport_op(grpc_exec_ctx *exec_ctx, grpc_transport *gt,
> > +                                 grpc_transport_op *op) {
> > +  grpc_chttp2_transport *t = (grpc_chttp2_transport *)gt;
> > +
> > +  lock(t);
> > +
> > +  /* Let's be overly cautious: don't change any state while we're parsing */
> > +  if (t->parsing_active) {
> > +    GPR_ASSERT(t->post_parsing_op == NULL);
> > +    t->post_parsing_op = gpr_malloc(sizeof(*op));
> > +    memcpy(t->post_parsing_op, op, sizeof(*op));
> > +  } else {
>
> Is it certain that this only gets called once wth parsing_active? Because
> otherwise the code looks inconsistent with the comment (we are changing the
> value of t->post_parsing_op after all).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/5560/files#r54977577>.
>